File: server/src/main/kotlin/cn/enaium/zensquare/controller/member/rank/MemberRankController.kt
/*
 * ZenSquare is an opensource forums
 *
 * Copyright (C) 2023  Enaium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package cn.enaium.zensquare.controller.member.rank

import cn.dev33.satoken.annotation.SaIgnore
import cn.enaium.zensquare.model.entity.Member
import cn.enaium.zensquare.model.entity.by
import cn.enaium.zensquare.repository.MemberRankRepository
import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

/**
 * Member rank controller
 *
 * @author Enaium
 */
@SaIgnore
@RestController
class MemberRankController(
    val memberRankRepository: MemberRankRepository
) {
    /**
     * Find member rank by post
     *
     * @param page page
     * @param size size
     * @return member rank
     */
    @GetMapping("/members/rank/thread/")
    fun findThreadRank(
        @RequestParam(defaultValue = "0") page: Int = 0,
        @RequestParam(defaultValue = "10") size: Int = 10
    ): Page<@FetchBy("DEFAULT_MEMBER_RANK") Member> {
        return memberRankRepository.findTop100OrderByThread(PageRequest.of(page, size))
    }

    /**
     * Find member rank by reply
     *
     * @param page page
     * @param size size
     * @return member rank
     */
    @GetMapping("/members/rank/reply/")
    fun findReplyRank(
        @RequestParam(defaultValue = "0") page: Int = 0,
        @RequestParam(defaultValue = "10") size: Int = 10
    ): Page<@FetchBy("DEFAULT_MEMBER_RANK") Member> {
        return memberRankRepository.findTop100OrderByReply(PageRequest.of(page, size))
    }

    companion object {
        val DEFAULT_MEMBER_RANK = newFetcher(Member::class).by {
            profile {
                avatar()
                nickname()
            }
            thread()
            reply()
            message()
        }
    }
}

File: website/src/__generated/services/MemberRankController.ts
import type { Executor } from '../';
import type { MemberDto } from '../model/dto';
import type { Page } from '../model/static';

export class MemberRankController {
    
    constructor(private executor: Executor) {}
    
    async findMessageRank(options: MemberRankControllerOptions['findMessageRank']): Promise<
        Page<MemberDto['MemberRankController/DEFAULT_MEMBER_RANK']>
    > {
        let _uri = '/members/rank/message/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.page;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'page='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.size;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'size='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Page<MemberDto['MemberRankController/DEFAULT_MEMBER_RANK']>
    }
    
    async findReplyRank(options: MemberRankControllerOptions['findReplyRank']): Promise<
        Page<MemberDto['MemberRankController/DEFAULT_MEMBER_RANK']>
    > {
        let _uri = '/members/rank/reply/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.page;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'page='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.size;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'size='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Page<MemberDto['MemberRankController/DEFAULT_MEMBER_RANK']>
    }
    
    async findThreadRank(options: MemberRankControllerOptions['findThreadRank']): Promise<
        Page<MemberDto['MemberRankController/DEFAULT_MEMBER_RANK']>
    > {
        let _uri = '/members/rank/thread/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.page;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'page='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.size;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'size='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Page<MemberDto['MemberRankController/DEFAULT_MEMBER_RANK']>
    }
}

export type MemberRankControllerOptions = {
    'findMessageRank': {readonly page?: number, readonly size?: number},
    'findReplyRank': {readonly page?: number, readonly size?: number},
    'findThreadRank': {readonly page?: number, readonly size?: number}
}